src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/LSPRazorProjectHost.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [AppliesTo("(DotNetCoreRazor & DotNetCoreRazorConfiguration) | ((DotNetCoreRazor | DotNetCoreWeb) & !DotNetCoreRazorConfiguration)")]
+    [Export(ExportContractNames.Scopes.UnconfiguredProject, typeof(IProjectDynamicLoadComponent))]
+    internal class LSPRazorProjectHost : IProjectDynamicLoadComponent
+    {
+        private readonly LSPEditorFeatureDetector _featureDetector;
+        private readonly List<Lazy<ILanguageClient, ILanguageClientMetadata>> _applicableLanguageClients;
+        private readonly Lazy<ILanguageClientBroker> _languageClientBroker;
+
+        [ImportingConstructor]
+        public LSPRazorProjectHost(
+            LSPEditorFeatureDetector featureDetector,
+            Lazy<ILanguageClientBroker> languageClientBroker,
+            [ImportMany] IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            if (featureDetector is null)
+            {
+                throw new ArgumentNullException(nameof(featureDetector));
+            }
+
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (languageClients is null)
+            {
+                throw new ArgumentNullException(nameof(languageClients));
+            }
+
+            _featureDetector = featureDetector;
+            _languageClientBroker = languageClientBroker;
+            _applicableLanguageClients = GetApplicableClients(languageClients);
+        }
+
+        public Task LoadAsync()
+        {
+            if (!_featureDetector.IsLSPEditorFeatureEnabled())
```

Comment:
I believe you have access to the project path at this point. We do [checks elsewhere](https://github.com/dotnet/aspnetcore-tooling/blob/e2ce8c9809acd3860e77940a246e722af95495b4/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorProjectChangePublisher.cs#L114) utilizing just the project path to ensure that the project itself can handle an LSP editor. It might be more consistent to utilize that instead of exposing the new `IsLSPEditorFeatureEnabled` flag.

Reply:
> I believe you have access to the project path at this point

By importing the `Workspace`?

> ensure that the project itself can handle an LSP editor

I checked the logic and all it does is check the capability of the project which we don't need to do here because this method won't get called if it doesn't have the right capabilities.

Reply:
> By importing the `Workspace`?

I think the UnconfiguredProject member has something on it.

> I checked the logic and all it does is check the capability of the project which we don't need to do here because this method won't get called if it doesn't have the right capabilities.

Doesnt it check additional C# capabilities too?

Reply:
Spoke offline. Couldn't use this because the logic inside `ProjectSupportsRazorLSPEditor` requires that the project already be loaded which is not possible at this stage.